doc/development/theming.rst

```diff
@@ -154,6 +160,157 @@ templating to put the color options into the stylesheet.  When a documentation
+        {{ meta.get("mykey") }}
+    {%- endif %}
+
+Note the check that ``meta`` is a dictionary ("mapping" in Jinja
+terminology) to ensure that using it in this way is valid.
+
+
+Defining custom template functions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Sometimes it is useful to define your own function in Python that you wish to
+then use in a template. For example, if you'd like to insert a template value
+with logic that depends on the user's configuration in the project, or if you'd
+like to include non-trivial checks and provide friendly error messages for
+incorrect configuration in the template.
+
+To define your own template function, you'll need to define two functions
+inside your module:
+
+* A **page context event handler** (or **registration**) function. This is
+  connected to the :class:`.Sphinx` application via an event callback.
+* A **template function** that you will use in your Jinja template.
+
+First, define the registration function, which accepts the arguments for
+:event:`html-page-context`.
+
+Within the registration function, define the template function that you'd like to use
+within Jinja. The template function should return a string or Python objects (lists,
+dictionaries) with strings inside that Jinja uses in the templating process
+
+.. note::
+
+    The template function will have access to all of the variables that
+    are passed to the registration function.
+
+At the end of the registration function, add the template function to the
+Sphinx application's context with ``context['template_func'] = template_func``.
+
+Finally, in your extension's ``setup()`` function, add your registration
+function as a callback for :event:`html-page-context`.
+
+.. code-block:: python
+
+   # The registration function
+    def setup_my_func(app, pagename, templatename, context, doctree):
+        # The template function
+        def my_func(mystring):
+            return "Your string is %s" % mystring
+        # Add it to the page's context
+        context['my_func'] = my_func
+
+    # Your extension's setup function
+    def setup(app):
+        app.connect("html-page-context", setup_my_func)
+
+Now, you will have access to this function in jinja like so:
+
+.. code-block:: jinja
+
+   <div>
+   {{ my_func("some string") }}
+   </div>
+
+
+Inject JavaScript based on user configuration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If your extension makes use of JavaScript, it can be useful to allow users
+to control its behavior using their Sphinx configuration. However, this can
+be difficult to do if your JavaScript comes in the form of a static library
+(which will not be built with Jinja).
+
+There are two ways to inject variables into the JavaScript space based on user
+configuration.
+
+First, you may append ``_t`` to the end of any static files included with your
+extension. This will cause Sphinx to process these files with the templating
+engine, allowing you to embed variables and control behavior.
```

Comment:
This behavior is not promised in public. And I'm still hesitating to support it officially (because it's not a "static" file!). So it should not be documented.

Reply:
Ah, so this is an implementation detail? I think we should either document this or remove this capability. If we do neither, [Hyrum's Law](https://www.hyrumslaw.com/) comes into play.

Reply:
Personally, I'd like to remove it ASAP. But it has behaved so since its beginning (maybe). And some extensions implicitly depend on the behavior. So we need to consider how to remove it safely.

Reply:
I believe it is documented in public here: https://www.sphinx-doc.org/en/master/theming.html#static-templates

that's where I found out about this feature, this PR is just trying to make it more discoverable. I guess I could remove this addition, but I feel like then it should be _removed_ from the documentation if it is explicitly not a desired feature.

Reply:
Okay, it's my wrong. I was thinking about `html_static_path`. But this is another topic. Please forget my comment above. As you quoted, files in the static directory of a theme are treated as a Jinja template and rendered on copying. That's a public feature.

What I described is `html_static_path`. It is also treated files under the specified directory as a Jinja template, but it is not official feature.